.doom.d/config.org
#+TITLE: Emacs config
#+AUTHOR: Daniel DeBruno (Renzix)
#+EMAIL: eragon9981@gmail.com

#+DESCRIPTION: My org mode config file
#+LANGUAGE: en
#+EXPORT_FILE_NAME: ../docs/index.html

# ReadTheOrg
#+SETUPFILE: https://fniessen.github.io/org-html-themes/setup/theme-readtheorg.setup

#+OPTIONS: num:nil

#+PROPERTY: header-args :results silent


Doom is litterally too good of me not to use it. There are alot of things i
would like to change to add to it tho.


Now that Im using doom all I have to do is worry about overriding some of the
defaults I dont like. If you havent already join the DOOM discord cuz its dope.
My old config can be found in my Dotfiles under =.config/emacs/config.org= and
my super old config is =.emacs.d/config.org=.

* Preferences
** Nice Settings

Default to emacs keybinds and keep evil installed so i can learn raw emacs. I
think I have used enough raw emacs bindings so I am going to switch back and do
it for certain modes ie org and terms. Also auto-revert-mode is for other
editors like ide's or nvim so i can use them to debug and possibly change 1-2
things.

#+NAME: evil
#+BEGIN_SRC emacs-lisp
(setq confirm-kill-emacs nil)
(global-auto-revert-mode t) ;; @NOTE(Renzix): For working with IDE's or other editors
;; Minibuffers in minibuffers
(setq enable-recursive-minibuffers t)
;; Mark stuff
(setq transient-mark-mode nil
      set-mark-command-repeat-pop t)
;; Use eww as the default browser cuz its nice
(setq browse-url-browser-function 'eww-browse-url
      search-whitespace-regexp ".*"
      isearch-lax-whitespace t)
#+END_SRC

** Visual
*** Themes

Themes r cool so lets do a different one for each day

#+NAME: Theme
#+BEGIN_SRC emacs-lisp
(load-theme 'doom-rouge t)
;; (when (display-graphic-p)
;;   (defvar renzix-weekday (format-time-string "%w"))
;;   (cond ((string= "0" renzix-weekday) ;; Sunday
;;          (load-theme 'doom-dracula t))
;;         ((string= "1" renzix-weekday) ;; Monday
;;          (load-theme 'doom-opera t))
;;         ((string= "2" renzix-weekday) ;; Tuesday
;;          (load-theme 'apropospriate-dark t))
;;         ((string= "3" renzix-weekday) ;; Wednesday
;;          (load-theme 'doom-molokai t))
;;         ((string= "4" renzix-weekday) ;; Thursday
;;          (load-theme 'doom-nord t))
;;         ((string= "5" renzix-weekday) ;; Friday
;;          (load-theme 'monokai t))
;;         ((string= "6" renzix-weekday) ;; Saturday
;;          (load-theme 'doom-one t))))
#+END_SRC

*** Highlighting

Highlight the current line

#+NAME: highlight-line
#+BEGIN_SRC emacs-lisp
(global-hl-line-mode)
#+END_SRC

Disable highlighting on marks and show marks globally

#+NAME: highlight-color
#+BEGIN_SRC emacs-lisp
;; (set-face-attribute 'region nil :background "#07B") ;; blue
(setq show-paren-priority 999)
(set-face-background 'region nil)
(global-visible-mark-mode t)
#+END_SRC

Highlights trailing whitespaceb, spaces before tabs, and characters over
line 100. Also highlight characters AT 80

#+NAME: highlight-whitespace
#+BEGIN_SRC emacs-lisp
;; @TODO(Renzix): Make this work in all themes?
(setq whitespace-style '(trailing lines-tail space-before-tab
                                  indentation space-after-tab)
      whitespace-line-column 81)
(add-hook! prog-mode-hook #'whitespace-mode)
;;(add-hook! prog-mode-hook (lambda () (highlight-regexp ".\{80\}\(.\)" 'hi-aquamarine "\\2"))) ;; @TODO(Renzix): Make this work as expected
#+END_SRC

*** Relative line Numbers

relative line numbers r lit

#+NAME: line-numbers
#+BEGIN_SRC emacs-lisp
(global-display-line-numbers-mode)
(setq-default display-line-numbers-type 'relative
              display-line-numbers-current-absolute t
              display-line-numbers-width 3
              display-line-numbers-widen t)
#+END_SRC

*** Font

#+BEGIN_SRC emacs-lisp
(setq doom-font (font-spec :family "Fira Code" :weight 'bold :size 14))
#+END_SRC

** Nice Functions
*** File Handling

Why isnt this already a function...

#+NAME: rename-file-and-buffer
#+BEGIN_SRC emacs-lisp

(defun my/rename-file-and-buffer ()
  "Renames current buffer and file it is visiting."
  (interactive)
  (let* ((name (buffer-name))
         (filename (buffer-file-name))
         (basename (file-name-nondirectory filename)))
    (if (not (and filename (file-exists-p filename)))
        (error "Buffer '%s' is not visiting a file!" name)
      (let ((new-name (read-file-name "New name: "
                                      (file-name-directory filename) basename nil basename)))
        (if (get-buffer new-name)
            (error "A buffer named '%s' already exists!" new-name)
          (rename-file filename new-name 1)
          (rename-buffer new-name)
          (set-visited-file-name new-name)
          (set-buffer-modified-p nil)
          (message "File '%s' successfully renamed to '%s'"
                   name (file-name-nondirectory new-name)))))))
#+END_SRC

This is the same problem as the function above. Emacs does not
close the buffer you have open if you delete the file so you might
accidently save it. Better to just call this function if the buffer
is open.

#+NAME: delete-file-and-buffer
#+BEGIN_SRC emacs-lisp
(defun my/delete-file-and-buffer ()
  "Kill the current buffer and deletes the file it is visiting."
  (interactive)
  (let ((filename (buffer-file-name)))
    (when filename
      (if (vc-backend filename)
          (vc-delete-file filename)
        (progn
          (delete-file filename)
          (message "Deleted file %s" filename)
          (kill-buffer))))))
#+END_SRC

Reads a string directly from a file then returns it as a string

#+NAME: get-string-from-file
#+BEGIN_SRC emacs-lisp
(defun my/get-string-from-file (filePath)
  "Return filePath's file content."
  (with-temp-buffer
    (insert-file-contents filePath)
    (buffer-string)))
#+END_SRC

*** Redefined Functions

This is C-a redefined to go to first nonwhitespace then if pressed
again go to actual start of line. Stolen from [[https://emacsredux.com/blog/2013/05/22/smarter-navigation-to-the-beginning-of-a-line/][here]] and from [[https://github.com/bbatsov/crux][crux]].

#+NAME: move-beginning-of-line
#+BEGIN_SRC emacs-lisp
(defun my/move-beginning-of-line (arg)
  "Move point back to indentation of beginning of line.

      Move point to the first non-whitespace character on this line.
      If point is already there, move to the beginning of the line.
      Effectively toggle between the first non-whitespace character and
      the beginning of the line.

      If ARG is not nil or 1, move forward ARG - 1 lines first.  If
      point reaches the beginning or end of the buffer, stop there."
  (interactive "^p")
  (setq arg (or arg 1))

  ;; Move lines first
  (when (/= arg 1)
    (let ((line-move-visual nil))
      (forward-line (1- arg))))

  (let ((orig-point (point)))
    (move-beginning-of-line 1)
    (when (= orig-point (point))
      (back-to-indentation))))
#+END_SRC

This is a function close to evil o and O which is pretty useful in
alot of scenerios. I have 0 idea why this isnt in emacs already. This
was also mainly stolen from [[https://emacsredux.com/blog/2013/03/26/smarter-open-line/][this]] and [[https://emacsredux.com/blog/2013/06/15/open-line-above/][this]] which was based off of [[https://github.com/bbatsov/crux][crux]] emacs
extension.

#+NAME: smart-open-line
#+BEGIN_SRC emacs-lisp
(defun my/smart-open-line (arg)
  "Insert an empty line after the current line.
      Position the cursor at its beginning, according to the current mode."
  (interactive "P")
  (if arg
      (progn
        (move-beginning-of-line nil)
        (newline-and-indent)
        (forward-line -1)
        (indent-according-to-mode))
    (progn
      (move-end-of-line nil)
      (newline-and-indent))))
#+END_SRC

*** Other Functions

I like indentation so lets indent smartly.

#+NAME: smart-indent
#+BEGIN_SRC emacs-lisp
(defun my/smart-indent ()
  "If a region is selected indent that.
If given ARG indent the current line.
Else indent the entire buffer."
  (interactive)
  (save-excursion
    (if (region-active-p) (progn
                            (indent-region (region-beginning) (region-end))
                            (message "Indenting region")))
    (if current-prefix-arg
        (progn
          (indent-region (line-beginning-position) (line-end-position))
          (message "Indenting line"))
      (progn
        (indent-region (point-min) (point-max))
        (message "Identing buffer")))))
#+END_SRC

#+NAME: select-line
#+BEGIN_SRC emacs-lisp
(defun my/select-line (&optional arg allow-extend)
  "Selects the current line and puts the cursor at the
start of the line.  A plain C-u puts the cursor at the
end of the line and any numbers puts the cursor at the
start and selects multiple lines(positive is down)"
  (interactive "p\np")
  (unless arg (setq arg 1))
  (when (zerop arg)
    (error "Cannot mark zero lines"))
  (cond ((and allow-extend
              (or (and (eq last-command this-command) (mark t))
                  (and transient-mark-mode mark-active)))
         (set-mark
          (save-excursion
            (goto-char (mark))
            (forward-line arg)
            (point))))
        (t
         (forward-line arg)
         (push-mark nil t t)
         (forward-line (* -1 arg)))))
#+END_SRC

#+NAME: select-word
#+BEGIN_SRC emacs-lisp
(defun my/select-word (&optional arg allow-extend)
  "Selects the current word and puts the cursor at the
start of the line.  A plain C-u puts the cursor at the
end of the line and any numbers puts the cursor at the
start and selects multiple lines(positive is down)"
  (interactive "p\np")
  (unless arg (setq arg 1))
  (when (zerop arg)
    (error "Cannot mark zero words"))
  (cond ((and allow-extend
              (or (and (eq last-command this-command) (mark t))
                  (and transient-mark-mode mark-active)))
         (set-mark
          (save-excursion
            (goto-char (mark))
            (forward-word arg)
            (point))))
        (t
         (forward-word arg)
         (push-mark nil t t)
         (backward-word arg))))  "Selects the current WORD"
#+END_SRC

Goes to a url with avy so I can open/goto it

#+NAME: my/avy-goto-url
#+BEGIN_SRC emacs-lisp
(defun my/avy-goto-url ()
  "Goto url with avy. Accepts http, https and file currently. "
  (interactive)
  (avy--generic-jump "https?s://\\|file://" nil nil))
#+END_SRC

* Packages
** Pretty Symbols

Here are some of my pretty symbols. alot are gonna be for perl6

#+NAME: pretty-symbols
#+BEGIN_SRC emacs-lisp
(after! org
  (setq +pretty-code-symbols
        (org-combine-plists +pretty-code-symbols
                            '(:pi "π"
                                  :tau "τ")))
  (set-pretty-symbols! 'perl6-mode
    :pi "pi"
    :tau "tau"))
#+END_SRC

** lsp

clang defaults to 2 spaces which is stupid and I am having other problems with
other stuff so I am going to disable it globally.

#+NAME: lsp
#+BEGIN_SRC emacs-lisp
(setq lsp-enable-indentation 'nil)
#+END_SRC

** evil

Emacs state in specific modes

#+NAME: evil
#+BEGIN_SRC emacs-lisp
;; Enable this for default emacs-state
;;(after! evil 
;;  (setq evil-default-state 'emacs))

;; Enable this for default evil stuff
(set-evil-initial-state! 'term-mode    'emacs)
(after! calc
  (set-evil-initial-state! 'calc-mode  'emacs))
(after! vterm
  (set-evil-initial-state! 'vterm-mode 'emacs))
(set-evil-initial-state! 'org-mode     'emacs)
(set-evil-initial-state! 'eshell-mode  'emacs)
(set-evil-initial-state! 'matrix-mode  'emacs)

(set-evil-initial-state! 'erc-mode     'emacs)
(set-evil-initial-state! 'circe-mode   'emacs)

#+END_SRC

*** Cursor

I like changing the cursor color based on the current mode i am in.

#+NAME: evil-cursor
#+BEGIN_SRC emacs-lisp
(setq evil-insert-state-cursor   '(bar "#FF00FF")
      evil-normal-state-cursor   '(box "#6666F6")
      evil-motion-state-cursor   '(hollow "#FFF500")
      evil-replace-state-cursor  '(hbar "#BF2222")
      evil-operator-state-cursor '(box "#FFA500")
      evil-visual-state-cursor   '(hollow "#FFFFFF")
      evil-emacs-state-cursor    '(box "#90EE90"))
(setq-default cursor-type 'bar)
(blink-cursor-mode 1)
#+END_SRC

** Org

I need to set plantuml path soo lets do that

#+NAME: plantuml-path
#+BEGIN_SRC emacs-lisp
(setq org-directory "~/Nextcloud/Documents"
      org-log-done 'timer
      org-plantuml-jar-path "/usr/share/java/plantuml/plantuml.jar")
;; Fix broken thing with emacs 27 where cursor is being hidden inside org mode
(custom-set-faces!
  '((hl-line solaire-hl-line-face org-indent
             outline-1 outline-2 outline-3 outline-4 outline-5 outline-6 outline-7 outline-8)
    :extend t))
#+END_SRC

** IRC

Doom uses circe. rcirc is also pretty nice also Im just gonna config this with
:pass. Note that for some weird reason this doesnt work on my gentoo install

#+NAME: irc
#+BEGIN_SRC emacs-lisp
(set-irc-server! "chat.freenode.net"
                 `(:tls t
                   :port 6697
                   :nick "Renzix"
                   :sasl-username "renzix"
                   :sasl-password ,(string-trim (my/get-string-from-file "~/.config/freenode-pass"))
                   :channels (:after-auth "#nixhub" "#emacs" "#kisslinux")))

(set-irc-server! "127.0.0.1"
                 `(
                   :port 6667
                   :nick "Renzix"
                   :nickserv-password ,(string-trim (my/get-string-from-file "~/.config/freenode-pass"))))
(after! circe
  (require 'circe-chanop)
  ;; Add some nice commands here - @TODO(Renzix)
  ;; Other config options
  (setq lui-time-stamp-position 'right-margin
        lui-fill-type nil
        lui-track-bar-behavior 'before-switch-to-buffer
        lui-flyspell-p t
        lui-flyspell-alist '((".*" "american")))
  (defun my-lui-setup ()
    (setq fringes-outside-margins t
          right-margin-width 5
          word-wrap t
          wrap-prefix "    "))
  (load "lui-logging" nil t)
  (enable-lui-logging-globally)
  (enable-lui-autopaste)
  (enable-lui-track-bar)
  (add-hook 'circe-channel-mode-hook 'enable-lui-autopaste)
  (add-hook 'lui-mode-hook 'my-lui-setup))
#+END_SRC

** Company

Trying out company-tabnine for autocompletions via deep learning and github

#+NAME: company-tabnine
#+BEGIN_SRC emacs-lisp
(after! company-mode
  (add-to-list 'company-backends #'company-tabnine))
#+END_SRC

** Helm

Helm is dope. I want to make it look better as I dont like the "minimal" look
that doom gives it. Thats only really good for ivy... I use my own version of
helm-M-x which adds a helpful-function action allowing me to get help without
=C-h f=.

#+NAME: helm
#+BEGIN_SRC emacs-lisp
(after! helm
  (defvar helm-source-emacs-commands
    (helm-build-sync-source "Emacs commands"
      :candidates (lambda ()
                    (let ((cmds))
                      (mapatoms
                       (lambda (elt) (when (commandp elt) (push elt cmds))))
                      cmds))
      :coerce #'intern-soft
      :action #'command-execute)
    "A simple helm source for Emacs commands.")
  (setq helm-mode-line-string t)
  (helm-add-action-to-source "Get Help" (lambda (candidate) (helpful-function candidate)) helm-source-emacs-commands 2))
(defun my/helm-M-x ()
  (interactive)
  "My emacs helm-M-x which also does Get Help so i dont have to C-h f"
  (helm :sources helm-source-emacs-commands))
#+END_SRC

** expand-region

Lets disable expand region word cuz i can just use M-d or other stuff

#+NAME: expand-region
#+BEGIN_SRC emacs-lisp
(after! expand-region
  (setq er/try-expand-list nil)
  (setq er/try-expand-list
        (append '(er/mark-method-call
                  er/mark-inside-quotes
                  er/mark-inside-pairs
                  er/mark-comment
                  er/mark-url
                  er/mark-email
                  er/mark-defun)
                er/try-expand-list)))
#+END_SRC

** elcord

Discord rich presence.

#+NAME: elcord
#+BEGIN_SRC emacs-lisp
(after! elcord
  (elcord-mode t))
#+END_SRC

** Piper

New thingy that allows me to do stuff with pipes interactively.

#+BEGIN_SRC emacs-lisp
(use-package! piper :commands piper)
#+END_SRC

** Twitch

Helm and Twitch integration.

#+BEGIN_SRC emacs-lisp
(use-package! helm-twitch
  :commands helm-twitch
  :init (setq twitch-api-oauth-token (string-trim (my/get-string-from-file "~/.config/twitch-oauth"))
              twitch-api-username "therenzix"))
#+END_SRC
** Notmuch

Email client for emacs. Uses isync to get email

#+NAME: notmuch
#+BEGIN_SRC emacs-lisp
(after! notmuch
  (setq notmuch-poll-script "getmail"
        +notmuch-sync-backend 'mbsync
        +notmuch-mail-folder "~/.mail"))
#+END_SRC

** Boon

#+NAME: boon
#+BEGIN_SRC emacs-lisp
(use-package! boon
  :config
  (boon-mode)
  (require 'boon-colemak))
#+END_SRC

* Keybindings
** Boon

#+BEGIN_SRC emacs-lisp
(map!
  "C-x C-c" 'save-buffers-kill-emacs
  (:map boon-command-map
   "p" #'projectile-command-map
   "f" #'isearch-forward
   "w" #'boon-qsearch-next
   "F" #'isearch-backward

.doom.d/config.el
(setq confirm-kill-emacs nil)
(global-auto-revert-mode t) ;; @NOTE(Renzix): For working with IDE's or other editors
;; Minibuffers in minibuffers
(setq enable-recursive-minibuffers t)
;; Mark stuff
(setq transient-mark-mode nil
      set-mark-command-repeat-pop t)
;; Use eww as the default browser cuz its nice
(setq browse-url-browser-function 'eww-browse-url
      search-whitespace-regexp ".*"
      isearch-lax-whitespace t)

(load-theme 'doom-rouge t)
;; (when (display-graphic-p)
;;   (defvar renzix-weekday (format-time-string "%w"))
;;   (cond ((string= "0" renzix-weekday) ;; Sunday
;;          (load-theme 'doom-dracula t))
;;         ((string= "1" renzix-weekday) ;; Monday
;;          (load-theme 'doom-opera t))
;;         ((string= "2" renzix-weekday) ;; Tuesday
;;          (load-theme 'apropospriate-dark t))
;;         ((string= "3" renzix-weekday) ;; Wednesday
;;          (load-theme 'doom-molokai t))
;;         ((string= "4" renzix-weekday) ;; Thursday
;;          (load-theme 'doom-nord t))
;;         ((string= "5" renzix-weekday) ;; Friday
;;          (load-theme 'monokai t))
;;         ((string= "6" renzix-weekday) ;; Saturday
;;          (load-theme 'doom-one t))))

(global-hl-line-mode)

;; (set-face-attribute 'region nil :background "#07B") ;; blue
(setq show-paren-priority 999)
(set-face-background 'region nil)
(global-visible-mark-mode t)

;; @TODO(Renzix): Make this work in all themes?
(setq whitespace-style '(trailing lines-tail space-before-tab
                                  indentation space-after-tab)
      whitespace-line-column 81)
(add-hook! prog-mode-hook #'whitespace-mode)
;;(add-hook! prog-mode-hook (lambda () (highlight-regexp ".\{80\}\(.\)" 'hi-aquamarine "\\2"))) ;; @TODO(Renzix): Make this work as expected

(global-display-line-numbers-mode)
(setq-default display-line-numbers-type 'relative
              display-line-numbers-current-absolute t
              display-line-numbers-width 3
              display-line-numbers-widen t)

(setq doom-font (font-spec :family "Fira Code" :weight 'bold :size 14))

(defun my/rename-file-and-buffer ()
  "Renames current buffer and file it is visiting."
  (interactive)
  (let* ((name (buffer-name))
         (filename (buffer-file-name))
         (basename (file-name-nondirectory filename)))
    (if (not (and filename (file-exists-p filename)))
        (error "Buffer '%s' is not visiting a file!" name)
      (let ((new-name (read-file-name "New name: "
                                      (file-name-directory filename) basename nil basename)))
        (if (get-buffer new-name)
            (error "A buffer named '%s' already exists!" new-name)
          (rename-file filename new-name 1)
          (rename-buffer new-name)
          (set-visited-file-name new-name)
          (set-buffer-modified-p nil)
          (message "File '%s' successfully renamed to '%s'"
                   name (file-name-nondirectory new-name)))))))

(defun my/delete-file-and-buffer ()
  "Kill the current buffer and deletes the file it is visiting."
  (interactive)
  (let ((filename (buffer-file-name)))
    (when filename
      (if (vc-backend filename)
          (vc-delete-file filename)
        (progn
          (delete-file filename)
          (message "Deleted file %s" filename)
          (kill-buffer))))))

(defun my/get-string-from-file (filePath)
  "Return filePath's file content."
  (with-temp-buffer
    (insert-file-contents filePath)
    (buffer-string)))

(defun my/move-beginning-of-line (arg)
  "Move point back to indentation of beginning of line.

      Move point to the first non-whitespace character on this line.
      If point is already there, move to the beginning of the line.
      Effectively toggle between the first non-whitespace character and
      the beginning of the line.

      If ARG is not nil or 1, move forward ARG - 1 lines first.  If
      point reaches the beginning or end of the buffer, stop there."
  (interactive "^p")
  (setq arg (or arg 1))

  ;; Move lines first
  (when (/= arg 1)
    (let ((line-move-visual nil))
      (forward-line (1- arg))))

  (let ((orig-point (point)))
    (move-beginning-of-line 1)
    (when (= orig-point (point))
      (back-to-indentation))))

(defun my/smart-open-line (arg)
  "Insert an empty line after the current line.
      Position the cursor at its beginning, according to the current mode."
  (interactive "P")
  (if arg
      (progn
        (move-beginning-of-line nil)
        (newline-and-indent)
        (forward-line -1)
        (indent-according-to-mode))
    (progn
      (move-end-of-line nil)
      (newline-and-indent))))

(defun my/smart-indent ()
  "If a region is selected indent that.
If given ARG indent the current line.
Else indent the entire buffer."
  (interactive)
  (save-excursion
    (if (region-active-p) (progn
                            (indent-region (region-beginning) (region-end))
                            (message "Indenting region")))
    (if current-prefix-arg
        (progn
          (indent-region (line-beginning-position) (line-end-position))
          (message "Indenting line"))
      (progn
        (indent-region (point-min) (point-max))
        (message "Identing buffer")))))

(defun my/select-line (&optional arg allow-extend)
  "Selects the current line and puts the cursor at the
start of the line.  A plain C-u puts the cursor at the
end of the line and any numbers puts the cursor at the
start and selects multiple lines(positive is down)"
  (interactive "p\np")
  (unless arg (setq arg 1))
  (when (zerop arg)
    (error "Cannot mark zero lines"))
  (cond ((and allow-extend
              (or (and (eq last-command this-command) (mark t))
                  (and transient-mark-mode mark-active)))
         (set-mark
          (save-excursion
            (goto-char (mark))
            (forward-line arg)
            (point))))
        (t
         (forward-line arg)
         (push-mark nil t t)
         (forward-line (* -1 arg)))))

(defun my/select-word (&optional arg allow-extend)
  "Selects the current word and puts the cursor at the
start of the line.  A plain C-u puts the cursor at the
end of the line and any numbers puts the cursor at the
start and selects multiple lines(positive is down)"
  (interactive "p\np")
  (unless arg (setq arg 1))
  (when (zerop arg)
    (error "Cannot mark zero words"))
  (cond ((and allow-extend
              (or (and (eq last-command this-command) (mark t))
                  (and transient-mark-mode mark-active)))
         (set-mark
          (save-excursion
            (goto-char (mark))
            (forward-word arg)
            (point))))
        (t
         (forward-word arg)
         (push-mark nil t t)
         (backward-word arg))))  "Selects the current WORD"

(defun my/avy-goto-url ()
  "Goto url with avy. Accepts http, https and file currently. "
  (interactive)
  (avy--generic-jump "https?s://\\|file://" nil nil))

(after! org
  (setq +pretty-code-symbols
        (org-combine-plists +pretty-code-symbols
                            '(:pi "π"
                                  :tau "τ")))
  (set-pretty-symbols! 'perl6-mode
    :pi "pi"
    :tau "tau"))

(setq lsp-enable-indentation 'nil)

;; Enable this for default emacs-state
;;(after! evil 
;;  (setq evil-default-state 'emacs))

;; Enable this for default evil stuff
(set-evil-initial-state! 'term-mode    'emacs)
(after! calc
  (set-evil-initial-state! 'calc-mode  'emacs))
(after! vterm
  (set-evil-initial-state! 'vterm-mode 'emacs))
(set-evil-initial-state! 'org-mode     'emacs)
(set-evil-initial-state! 'eshell-mode  'emacs)
(set-evil-initial-state! 'matrix-mode  'emacs)

(set-evil-initial-state! 'erc-mode     'emacs)
(set-evil-initial-state! 'circe-mode   'emacs)

(setq evil-insert-state-cursor   '(bar "#FF00FF")
      evil-normal-state-cursor   '(box "#6666F6")
      evil-motion-state-cursor   '(hollow "#FFF500")
      evil-replace-state-cursor  '(hbar "#BF2222")
      evil-operator-state-cursor '(box "#FFA500")
      evil-visual-state-cursor   '(hollow "#FFFFFF")
      evil-emacs-state-cursor    '(box "#90EE90"))
(setq-default cursor-type 'bar)
(blink-cursor-mode 1)

(setq org-directory "~/Nextcloud/Documents"
      org-log-done 'timer
      org-plantuml-jar-path "/usr/share/java/plantuml/plantuml.jar")
;; Fix broken thing with emacs 27 where cursor is being hidden inside org mode
(custom-set-faces!
  '((hl-line solaire-hl-line-face org-indent
             outline-1 outline-2 outline-3 outline-4 outline-5 outline-6 outline-7 outline-8)
    :extend t))

(set-irc-server! "chat.freenode.net"
                 `(:tls t
                   :port 6697
                   :nick "Renzix"
                   :sasl-username "renzix"
                   :sasl-password ,(string-trim (my/get-string-from-file "~/.config/freenode-pass"))
                   :channels (:after-auth "#nixhub" "#emacs" "#kisslinux")))

(set-irc-server! "127.0.0.1"
                 `(
                   :port 6667
                   :nick "Renzix"
                   :nickserv-password ,(string-trim (my/get-string-from-file "~/.config/freenode-pass"))))
(after! circe
  (require 'circe-chanop)
  ;; Add some nice commands here - @TODO(Renzix)
  ;; Other config options
  (setq lui-time-stamp-position 'right-margin
        lui-fill-type nil
        lui-track-bar-behavior 'before-switch-to-buffer
        lui-flyspell-p t
        lui-flyspell-alist '((".*" "american")))
  (defun my-lui-setup ()
    (setq fringes-outside-margins t
          right-margin-width 5
          word-wrap t
          wrap-prefix "    "))
  (load "lui-logging" nil t)
  (enable-lui-logging-globally)
  (enable-lui-autopaste)
  (enable-lui-track-bar)
  (add-hook 'circe-channel-mode-hook 'enable-lui-autopaste)
  (add-hook 'lui-mode-hook 'my-lui-setup))

(after! company-mode
  (add-to-list 'company-backends #'company-tabnine))

(after! helm
  (defvar helm-source-emacs-commands
    (helm-build-sync-source "Emacs commands"
      :candidates (lambda ()
                    (let ((cmds))
                      (mapatoms
                       (lambda (elt) (when (commandp elt) (push elt cmds))))
                      cmds))
      :coerce #'intern-soft
      :action #'command-execute)
    "A simple helm source for Emacs commands.")
  (setq helm-mode-line-string t)
  (helm-add-action-to-source "Get Help" (lambda (candidate) (helpful-function candidate)) helm-source-emacs-commands 2))
(defun my/helm-M-x ()
  (interactive)
  "My emacs helm-M-x which also does Get Help so i dont have to C-h f"
  (helm :sources helm-source-emacs-commands))

(after! expand-region
  (setq er/try-expand-list nil)
  (setq er/try-expand-list
        (append '(er/mark-method-call
                  er/mark-inside-quotes
                  er/mark-inside-pairs
                  er/mark-comment
                  er/mark-url
                  er/mark-email
                  er/mark-defun)
                er/try-expand-list)))

(after! elcord
  (elcord-mode t))

(use-package! piper :commands piper)

(use-package! helm-twitch
  :commands helm-twitch
  :init (setq twitch-api-oauth-token (string-trim (my/get-string-from-file "~/.config/twitch-oauth"))
              twitch-api-username "therenzix"))

(after! notmuch
  (setq notmuch-poll-script "getmail"
        +notmuch-sync-backend 'mbsync
        +notmuch-mail-folder "~/.mail"))

(use-package! boon
  :config
  (boon-mode)
  (require 'boon-colemak))

(map!
  "C-x C-c" 'save-buffers-kill-emacs
  (:map boon-command-map
   "p" #'projectile-command-map
   "f" #'isearch-forward
   "w" #'boon-qsearch-next
   "F" #'isearch-backward
   "W" #'boon-qsearch-previous
   "h" #'helm-find-files
   "H" #'helm-mini
   "/" #'undo)
  (:map projectile-command-map
   "s" #'helm-projectile-rg))

(setq doom-leader-alt-key "C-c"
      doom-localleader-alt-key "C-c l")
(map!
 :nve  "C-x t"   #'+eshell/here
 :nve  "C-x C-t" #'+vterm/here
 :nvei "C-<tab>" #'+treemacs/toggle
 (:leader ;; Make this open guix if guixsd???
  :desc "System Packages" "P" #'helm-system-packages
  :desc "Open Link" "o l" #'my/avy-goto-url)
 (:map smerge-mode-map ;; cuz C-c ^ is awful C-c l m is much better
  :localleader
  :desc "Merge" "m" #'smerge-command-prefix))

(map!
 :e "C-x C-k" #'kill-this-buffer
 :e "C-a"     #'my/move-beginning-of-line
 :e "C-e"     #'end-of-line
 :e "C-j"     #'avy-goto-char-2
 :e "M-l"     #'my/select-line
 :e "M-e"     #'my/select-word
 :e "M-a"     #'er/mark-inside-pairs
 :e "M-h"     #'mark-defun
 :e "M-y"     #'yank-pop
 :e "C-\\"    #'er/expand-region
 :e "C-="     #'my/smart-indent
 :e "C-o"     #'my/smart-open-line
 :e "C-M-u"   #'sp-backward-up-sexp
 :e "C-u"     #'universal-argument ;; Doom rebinds this idk why
 (:map override
  :e "C-;"   #'my/helm-M-x ;; I dont know if i shoud have this or not
  :e "C-:"   #'evil-ex
  :e "C-'"   #'helm-find-files
  :e "C-\""  #'helm-mini
  :e "M-'"   #'+helm/projectile-find-file
  :e "M-\""  #'+helm/project-search))

(map!
 :nv "Q"     (kbd "@q")
 :nv "g ="   #'my/smart-indent
 (:map override
  :nvm ";"   #'my/helm-M-x
  :nvm "|"   #'helm-mini
  :nvm "s"   #'helm-find-files
  :nvm "S"   #'+helm/projectile-find-file
  :nv  "U"   #'undo-tree-visualize ;; in vi U is undo line changes so you can undo the undo
  :nvm "\\"  #'+helm/project-search))
;; @NOTE(Renzix) that I made these from evil functions to emacs function
;; for more compatibility and to make sure it works as expected.
(after! evil
  (evil-ex-define-cmd "cfg" (lambda! (find-file "~/Dotfiles/.doom.d/config.org")))
  (evil-ex-define-cmd "conf[ig]" (lambda! (find-file "~/Dotfiles/.doom.d/config.org")))
  (evil-ex-define-cmd "pkg" (lambda! (find-file "~/Dotfiles/.doom.d/packages.el")))
  (evil-ex-define-cmd "pack[age]" (lambda! (find-file "~/Dotfiles/.doom.d/packages.el")))
  (evil-ex-define-cmd "init" (lambda! (find-file "~/Dotfiles/.doom.d/init.el")))
  (evil-ex-define-cmd "blog" (lambda! (find-file "~/Blog/blog.org")))
  (evil-ex-define-cmd "a[genda]" #'org-agenda)
  (evil-ex-define-cmd "q[uit]" #'delete-window)
  (evil-ex-define-cmd "bd" #'kill-this-buffer))

(map! (:map helm-map
        "C-m" #'helm-maybe-exit-minibuffer
        ;;"<return>" #'helm-maybe-exit-minibuffer
        "C-j"   #'helm-execute-persistent-action
        "C-<return>" #'helm-execute-persistent-action
        "C-i"   #'helm-select-action ;; Terminal Tab
        "<tab>" #'helm-select-action))

(map! :map org-mode-map
      :e "C-c e" #'org-edit-src-code) ;; @TODO(Renzix): Make this work on C-c '

(map! :map vterm-mode-map
      :e "C-a"   #'vterm--self-insert
      :e "C-h"   #'vterm--self-insert
      :e "C-j"   #'vterm--self-insert
      :e "C-k"   #'vterm--self-insert
      :e "C-l"   #'vterm--self-insert
      :e "C-e"   #'vterm--self-insert
      :e "C-r"   #'vterm--self-insert
      :e "C-s"   #'vterm--self-insert
      :e "C-u"   #'vterm--self-insert
      :e "C-x u" #'vterm--self-insert
      :e "C-/"   #'vterm--self-insert
      :e "C-y"   #'vterm--self-insert
      :e "M-y"   #'vterm--self-insert
      :e "RET"   #'vterm--self-insert)
